### ISSUE TITLE
Understanding of Eq.2 in paper. 

### ISSUE THREAD

**Issue by yanqswhu** [princeton-vl/SEA-RAFT#22]
Eq. 2 in SEA-RAFT is 

$$
-log[MixLap(x,\alpha,\beta_1,\beta_2,\mu)]
$$

where

$$
MixLap(x,\alpha,\beta_1,\beta_2,\mu)=\alpha\frac{e^{-\frac{|x-\mu|}{\beta_1}}}{2\beta_1} + (1-\alpha)\frac{e^{-\frac{|x-\mu|}{\beta_2}}}{2\beta_2}
$$

replace $\alpha=\frac{exp(\alpha_1)}{exp(\alpha_1)+exp(\alpha_2)}$

$$
MixLap(x,\alpha,\beta_1,\beta_2,\mu)=\frac{exp(\alpha_1)}{exp(\alpha_1)+exp(\alpha_2)}\frac{e^{-\frac{|x-\mu|}{\beta_1}}}{2\beta_1}+\frac{exp(\alpha_2)}{exp(\alpha_1)+exp(\alpha_2)}\frac{e^{-\frac{|x-\mu|}{\beta_2}}}{2\beta_2}
$$

rewrite as follows

$$
MixLap(x,\alpha,\beta_1,\beta_2,\mu)=\frac{exp(\alpha_1)\frac{e^{-\frac{|x-\mu|}{\beta_1}}}{2\beta_1}+exp(\alpha_2)\frac{e^{-\frac{|x-\mu|}{\beta_2}}}{2\beta_2}}{exp(\alpha_1)+exp(\alpha_2)}
$$

therefore, 

$$
-log[MixLap(x,\alpha,\beta_1,\beta_2,\mu)]=-(log[{exp(\alpha_1)\frac{e^{-\frac{|x-\mu|}{\beta_1}}}{2\beta_1}+exp(\alpha_2)\frac{e^{-\frac{|x-\mu|}{\beta_2}}}{2\beta_2}}]-log[{exp(\alpha_1)+exp(\alpha_2)}])
$$

remove the negtive

$$
-log[MixLap(x,\alpha,\beta_1,\beta_2,\mu)]=log[{exp(\alpha_1)+exp(\alpha_2)}]-log[{exp(\alpha_1)\frac{e^{-\frac{|x-\mu|}{\beta_1}}}{2\beta_1}+exp(\alpha_2)\frac{e^{-\frac{|x-\mu|}{\beta_2}}}{2\beta_2}}]
$$

----

Notes in code `raft.py Line145-Line156` , where $\alpha,\beta,\mu$ is replace by a, b, u：

obtain  $log(\beta)$ with shape $(N,2,H,W)$
```python
raw_b = info_predictions[i][:, 2:]
log_b = torch.zeros_like(raw_b)
# Large b Component                
log_b[:, 0] = torch.clamp(raw_b[:, 0], min=0, max=var_max)
# Small b Component
log_b[:, 1] = torch.clamp(raw_b[:, 1], min=var_min, max=0)
```

obtain $\alpha$ with shape  $(N,2,H,W)$, which can be splited into $\alpha_1,\alpha_2$ on `dim=1`

```python
weight = info_predictions[i][:, :2]
```

obtain $|x-\mu|/\beta$  with shape $(N,2,H,W)$
```python
# |x-u|*exp[-log(b)]
# |x-u|/b
term2 = ((flow_gt - flow_predictions[i]).abs().unsqueeze(2)) * (torch.exp(-log_b).unsqueeze(1))
```

obtain $\alpha-log(2)-log(\beta)$ with shape  $(N,2,H,W)$
```python
# term1: [N, m, H, W]
# a-log(2)-log(b)
term1 = weight - math.log(2) - log_b
```

obtain nf_loss= $-log[MixLap(x,\alpha,\beta_1,\beta_2,\mu)]$ with shape  $(N,1,H,W)$
```python
nf_loss = torch.logsumexp(weight, dim=1, keepdim=True) - torch.logsumexp(term1.unsqueeze(1) - term2, dim=2)
```
where `torch.logsumexp(weight, dim=1, keepdim=True)` = $log[exp(\alpha_1)+exp(\alpha_2)]$,
`torch.logsumexp(term1.unsqueeze(1) - term2, dim=2)` = $log[exp(\alpha_1-log(2)-log(\beta_1)-|x-\mu|/\beta_1)+exp(\alpha_2-log(2)-log(\beta_2)-|x-\mu|/\beta_2)]$
$=log[exp (\alpha_1) \frac{1}{2\beta_1} e^{-\frac{|x-\mu|}{\beta_1}}+exp (\alpha_2) \frac{1}{2\beta_2} exp^{-\frac{|x-\mu|}{\beta_2}}$

Hope this helps guys read the code！

**Comment by MemorySlices** [comment]
Thanks for the explanation!

**Comment by YangHai-1218** [comment]
Very good explanation. Thanks!

**Comment by infinity1096** [comment]
This greatly helps to understand the code. Thanks!

**Comment by Neoyning** [comment]
Hi, I debug the code in `custom.py`, which show the size of nf_loss at  [https://github.com/princeton-vl/SEA-RAFT/blob/886fb094fe21d4fa5ff675da18362b27b023ccc3/core/raft.py#L156](url)

nf_loss has shape 
```
torch.size([1,2,218,512]) #(N,2,H,W)?
```

,which different from answer mentioned above $(N,1,H,W)$ , is this indicate nf_loss is 2 dimension for each pixel?

The code for `term2` in [https://github.com/princeton-vl/SEA-RAFT/blob/886fb094fe21d4fa5ff675da18362b27b023ccc3/core/raft.py#L153](url)

```
term2 = ((flow_gt - flow_predictions[i]).abs().unsqueeze(2)) * (torch.exp(-log_b).unsqueeze(1))
```

where `((flow_gt - flow_predictions[i]).abs().unsqueeze(2))` has shape [1,**2,1**,218,512] $(N,2,1,H,W)$ due to the 'unsqueeze(2)' operation

and `(torch.exp(-log_b).unsqueeze(1))` has shape [1,**2,1**,218,512] $(N,1,2,H,W)$ also because of the 'unsqueeze(1)' (why 'unsqueeze(1)' instead of 'unsqueeze(2)' ?)

By boardcasting law, `term2` then has shape [1,2,2,218,512] $(N,2,2,H,W)$ , which is also different from the answer above $(N,2,H,W)$.

A bit confusing, thanks for further explanation if possible!
